// vast-player/lib/players/flashTester.js
'use strict';
var swfobject = require('./myswf');
var FLASH_TEST = 'vpaid_video_flash_tester';
var FLASH_TEST_EL = 'vpaid_video_flash_tester_el';
var JSFlashBridge = require('./jsFlashBridge').JSFlashBridge;
var utils = require('./utils');
var MultipleValuesRegistry = require('./registry').MultipleValuesRegistry;
function FlashTester(parent, swfConfig)  {
  this.parentEl = utils.createElementWithID(parent, FLASH_TEST_EL); // some browsers create global variables using the element id http://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-variables
        utils.hideFlashEl(this.parentEl);
		 
        var params = {};
        params.movie = swfConfig.data;
		
       // params.FlashVars = `flashid=${FLASH_TEST_EL}&handler=${JSFlashBridge.VPAID_FLASH_HANDLER}`;
	    params.FlashVars = 'flashid='+FLASH_TEST_EL+'&handler='+JSFlashBridge.VPAID_FLASH_HANDLER; 
		params.allowScriptAccess = 'always';
		this.el = swfobject.createSWF(swfConfig, params, FLASH_TEST_EL);
		this._handlers = new MultipleValuesRegistry();
		
		this._isSupported = false;
		var self = this;
        if (this.el) {
		utils.hideFlashEl(this.el);
		
		this._flash = new JSFlashBridge(this.el, swfConfig.data, FLASH_TEST_EL, swfConfig.width, swfConfig.height, function(){
		var  support = true;
		self._isSupported = support;
		alert(support);
		self._handlers.get('change').forEach(function(callback){
		alert(callback);
                    setTimeout(function(){
                        callback('change', support);
                    }, 0);
                });
            });
		}

    };
FlashTester.prototype.isSupported=function() {
        return this._isSupported;
    };
FlashTester.prototype.on=function(eventName, callback) {
        alert("доавить");
        this._handlers.add(eventName, callback);
    };

var createFlashTester = function (el, swfConfig) {
    if (!window[FLASH_TEST]) {
        window[FLASH_TEST] = new FlashTester(el, swfConfig);
    }
    return window[FLASH_TEST];
};
module.exports.createFlashTester = createFlashTester;
