// vast-player/lib/players/JavaScriptVPAID.js
'use strict';

var inherits = require('util').inherits;
var VPAID = require('./VPAID');
var LiePromise = require('lie');
var EVENTS = require('../enums/VPAID_EVENTS');
var isDesktop = require('../environment').isDesktop;
var VPAIDVersion = require('../VPAIDVersion');

function JavaScriptVPAID() {
    VPAID.apply(this, arguments); // call super()
    this.playDelay=1;
	this.playClean=0;
	this.mypaused=1;
    this.frame = null;
}
inherits(JavaScriptVPAID, VPAID);

JavaScriptVPAID.prototype.load = function load(mediaFiles, parameters) {
    var self = this;
    var uri = mediaFiles[0].uri;
    var bitrate = mediaFiles[0].bitrate;

    return new LiePromise(function loadCreative(resolve, reject) {
        var iframe = document.createElement('iframe');
		iframe.scrolling="no";
        var script = document.createElement('script');
        var video = document.createElement('video');
        function checkMyOPaused(api){
		} 
        function cleanup(reason) {
		    self.playClean=1;
		    try{
            self.container.removeChild(iframe);
			}catch(e){
			
			}
            self.frame = null;
            self.api = null;
           
            if (reason) {

                reject(reason);
            }
			
        }
		function setCheckPlayedTime(cntS){
		if(self.playClean) return;
		//console.log(["222+ ",cntS]);   
		checkMyOPaused(self.vpaApi);
		
		if(cntS>0){
		setTimeout(function(){
		setCheckPlayedTime((cntS-1))
		}, 5000);
		return;
		}
		
		
		self.stopAd();
		}
        function setCheckLoadedTime(cnt){
		if(self.playClean) return;
		if(!self.playDelay) return;
			//console.log("vpaid "+cnt);  
		if(cnt>0){
		setTimeout(function(){
		setCheckLoadedTime((cnt-1))
		}, 1000);
		return;
		}
		cleanup(new Error("vpaid не загрузился в течении 18 сек"));  
		}
		
		setCheckLoadedTime(14);
		
		
        iframe.src = 'about:blank';
        iframe.style.width = '100%';
        iframe.style.height = '100%';
        iframe.style.display = 'block';
        iframe.style.opacity = '0';
        iframe.style.border = 'none';

        video.setAttribute('webkit-playsinline', 'true');
        video.style.display = 'block';
        video.style.width = '100%';
        video.style.height = '100%';
        video.style.objectFit = 'contain';
        
        self.container.appendChild(iframe);
        // Opening the iframe document for writing causes it to inherit its parent's location
        iframe.contentWindow.document.open();
        iframe.contentWindow.document.close();

        iframe.contentWindow.document.body.style.margin = '0';
        self.frame = iframe;
         script.src = uri;
		if(self.id1==-4){
		//script.src=uri.replace(/https\:\/\//,'//');
        //console.log(["третья урна",self.id1,uri]);
		}
		//console.log(["третья тема",self.id1,script.src]); 
		
		console.log(["третья стадия",self.id1,script.src]); 
        script.onload = function onload() {
		
            var vpaid = iframe.contentWindow.getVPAIDAd();
			if(self.id1==33 || self.id1 == 34){
					video.style.display="none"; 
		            }	
		    var position = iframe.getBoundingClientRect();
            var slot = iframe.contentWindow.document.body;
            var version = self.vpaidVersion = new VPAIDVersion(vpaid.handshakeVersion('2.0'));
            
            function resizeAd() {
                var position = iframe.getBoundingClientRect();
                self.resizeAd(position.width, position.height, 'normal');
            }

            if (version.major > 2) {
                return reject(new Error('VPAID version ' + version + ' is not supported.'));
            }

            iframe.contentWindow.addEventListener('resize', resizeAd, false);
            
            EVENTS.forEach(function subscribe(event) {
                return vpaid.subscribe(function handle(/*...args*/) {
                    var args = new Array(arguments.length);
                    var length = arguments.length;
					
                    while (length--) { args[length] = arguments[length]; }
					
                    return self.emit.apply(self, [event].concat(args));
                }, event);
            });
            
				    
            self.once(EVENTS.AdLoaded, function onAdLoaded() {
              
			    self.playDelay=0;
                iframe.style.opacity = '1';
                self.api = vpaid;
				if(self.id1==31 || self.id1 == 32){
				
				}else{
				//setCheckPlayedTime(20);
				}
                resolve(self);
            });

            self.once(EVENTS.AdError, function onAdError(reason) {
			
		        cleanup(new Error(reason));
            });

            self.once(EVENTS.AdStopped, function (){
			cleanup();
			});
			self.vpaApi=vpaid;
			if(self.id1==31 || self.id1 == 32){
			self.once(EVENTS.AdPlaying, function onAdStart() {
			setCheckPlayedTime(12);
			});
			}
			
            vpaid.initAd(
                position.width,
                position.height,
                'normal',
                bitrate,
                { AdParameters: parameters },
                { slot: slot, videoSlot: video, videoSlotCanAutoPlay: isDesktop }
            );
        };
        script.onerror = function onerror() {
        cleanup(new Error('Failed to load MediaFile [' + uri + '].'));
        };

        iframe.contentWindow.document.body.appendChild(video);
        iframe.contentWindow.document.head.appendChild(script);
    });
};

module.exports = JavaScriptVPAID;
