// vast-player/lib/players/utils.js
'use strict';
var LetCount = -1;
exports.unique= function unique(prefix) {
   
    return function (){
        return prefix+'_'+(++LetCount);
    };
}
function noop() {
}

exports.noop = noop;

function callbackTimeout(timer, onSuccess, onTimeout) {
   
    var timeout = setTimeout(function() {
   
        onSuccess = noop;
        onTimeout();

    }, timer);
    
    return function () {
        clearTimeout(timeout);
		console.log( arguments);
        onSuccess.apply(this, arguments);
    };
}
exports.callbackTimeout = callbackTimeout;

exports.createElementWithID = function createElementWithID(parent, id, cleanContent) {
cleanContent=cleanContent||false;
    var nEl = document.createElement('div');
    nEl.id = id;
    if (cleanContent) {
        parent.innerHTML = '';
    }
    parent.appendChild(nEl);
    return nEl;
}

exports.isPositiveInt = function isPositiveInt(newVal, oldVal) {
    return !isNaN(parseFloat(newVal)) && isFinite(newVal) && newVal > 0 ? newVal : oldVal;
}

var endsWith = (function () {
    if (String.prototype.endsWith) return String.prototype.endsWith;
    return function endsWith (searchString, position) {
        var subjectString = this.toString();
        if (position === undefined || position > subjectString.length) {
            position = subjectString.length;
        }
        position -= searchString.length;
        var lastIndex = subjectString.indexOf(searchString, position);
        return lastIndex !== -1 && lastIndex === position;
    };
})();

exports.stringEndsWith=function stringEndsWith(string, search) {
    return endsWith.call(string, search);
}

exports.hideFlashEl = function hideFlashEl(el) {
    // can't use display none or visibility none because will block flash in some browsers
    el.style.position = 'absolute';
    el.style.left = '-1px';
    el.style.top = '-1px';
    el.style.width = '1px';
    el.style.height = '1px';
}
