// vast-player/lib/players/VPAIDFLASHClient.js
'use strict';
 	    
var FLASH_TEST_EL = 'vpaid_video_flash_tester_el';
var swfobject = require('./myswf');
var noop = require('./utils').noop;
var isPositiveInt = require('./utils').isPositiveInt;
var uniqueVPAID = require('./utils').unique('vpaid');
var callbackTimeout = require('./utils').callbackTimeout;
var createElementWithID = require('./utils').createElementWithID;
var JSFlashBridge = require('./jsFlashBridge').JSFlashBridge;
var VPAIDAdUnit = require('./VPAIDAdUnit').VPAIDAdUnit;
//var createFlashTester = require('./flashTester').createFlashTester;
var ERROR = 'error';
var FLASH_VERSION = '10.1.0';
var flashTester=true;
 	   
function VPAIDFLASHClient(elementP, callback, swfConfig, params , vpaidOptions) {
 	    var me = this;
        this._vpaidParentEl = vpaidParentEl;
		this._flashID = uniqueVPAID();
		this._destroyed = false;
		this._playerLoading = 0;
	
        callback = callback || noop;
	   
		swfConfig.width = isPositiveInt(swfConfig.width, 500);
        swfConfig.height = isPositiveInt(swfConfig.height, 350);

	    if(!VPAIDFLASHClient.isSupported()) {
	    
         return onError('user don\'t support flash or doesn\'t have the minimum required version of flash ' + FLASH_VERSION);
        }
		
        params.movie = swfConfig.data;
		var vpaidParentEl=document.createElement("DIV"); 
		document.body.appendChild(vpaidParentEl);
        vpaidParentEl.style.display='block!important';
		var parentEl = createElementWithID(vpaidParentEl, FLASH_TEST_EL); // some browsers create global variables
		var self = this;
	    params.FlashVars = 'flashid='+FLASH_TEST_EL+'&handler='+JSFlashBridge.VPAID_FLASH_HANDLER+'&debug='+vpaidOptions.debug+'salign='+params.salign; 
		params.FlashVars = 'flashid='+FLASH_TEST_EL+'&handler='+JSFlashBridge.VPAID_FLASH_HANDLER+'&debug=false'; 
		params.allowScriptAccess = 'always';
		//console.log(11112);
	    this.el = swfobject.createSWF(swfConfig, params, FLASH_TEST_EL);
		//console.log(11112);
		if(!this.el){
		 document.body.removeChild(vpaidParentEl);
		 return onError('user don\'t support flash or doesn\'t have the minimum required version of flash ' + FLASH_VERSION);
		}
					this._flash = new JSFlashBridge(this.el, swfConfig.data, FLASH_TEST_EL, swfConfig.width, swfConfig.width, function(){
					self._playerLoading = 1;
			        callback();
		            });
					
					
					
	    function setCheckLoadedTime(cnt){
		if(self._playerLoading) return;
		if(self._destroyed) return;
		console.log(["таймер load",cnt]);
		if(cnt>0){
		setTimeout(function(){
		setCheckLoadedTime((cnt-1))
		}, 1000);
		return;
		}
         self.destroy();
	     document.body.removeChild(vpaidParentEl);
		 return onError("flash 10 сек");
		 //reject("flash 10 сек");
		//cleanup(new Error("vpaid не загрузился в течении 18 сек"));  
		}
		
		setCheckLoadedTime(10);			
		
		return;
		
		console.log(11115);
			this._flash = new JSFlashBridge(this.el, swfConfig.data, FLASH_TEST_EL, swfConfig.width, swfConfig.width, function(){
			console.log(11116);
                // var adURL = 'http://cdn.innovid.com/2.62.8110/platform/vpaid/VPAIDIRollPackage.swf?configURL=http%3A%2F%2Fstatic.innovid.com%2Firoll%2Fconfig%2F1hl7lc.xml%3Fcb%3D787766d7-ebab-3656-c24f-0ddebab645e9&secure=false';
                // var adURL = 'VPAIDIRollPackage.swf?configURL=http%3A%2F%2Fstatic.innovid.com%2Firoll%2Fconfig%2F1hl7lc.xml%3Fcb%3D787766d7-ebab-3656-c24f-0ddebab645e9&secure=false';
                // var adURL = 'TestAd.swf';
                var adURL = 'http://cdn-sys.brainient.com/flash/v6/select846.swf?video_id=a3f30b8e-2ad8-4123-bc58-42fccb3e48cd&user_id=1228&tzone=&settings=json&settingsPath=http://cdn-tags.brainient.com/1228/a3f30b8e-2ad8-4123-bc58-42fccb3e48cd/config.json';
                // var adURL = 'http://shim.btrll.com/shim/20150715.85603_master/Scout.swf?asset_64=aHR0cDovL2NhY2hlLmJ0cmxsLmNvbS9wcm9kdWN0L3Rlc3QvdmFzdF93cmFwcGVyL2JyLXZhc3Rfd3JhcHBlci54bWw&vid_click_url=&h_64=YnJ4c2Vydi0yMi5idHJsbC5jb20&e=p&config_url_64=&type=VAST_TAG&vh_64=bWhleHQtMjIuYnRybGwuY29t&p=6834995&s=3863356&l=28043&ic=51223&ii=6594&x=TbBvLqwwDICcRVsPZkAABtiwAAyBcAOvM8AAAAAABhtJT2o-vMJQ&cx=&dn=&hidefb=true&iq=t&adc=false&si=&t=33&apep=0.03&hbp=0.01&epx=&ps=0.0&view=vast2&woid=____________________________________';
//var adURL = "https://cdn.webturn.ru/-/OtherVAST/LamodaVASTFeb17_1.swf?"+Math.random();
//var adURL="https://imasdk.googleapis.com/flash/sdkloader/vpaid2video.swf?adTagUrl=embedded&embedAdsResponse=1";
               self.loadAdUnit(adURL,function (error, adUnit){
			   adUnit.handshakeVersion('2.0', initAd);
               adUnit.on('AdLoaded', startAd);
                function initAd(err, result) {
                        console.log('handShake', err, result);
                        adUnit.initAd(swfConfig.width, swfConfig.height, 'normal', -1, '', '', function (err) {
                        });
                }
				function startAd(err, result) {
				       // elementP.style.display='block';
						document.getElementById("mycontoller").style.display="none";
						//elementP.innerHTML="df;gk;d rwwerwerwrwreflkg;dflgk;dflgk;dflgkdf;glkgdf;lgk";
						//elementP.style.color='#FFFFFF';
					//vpaidParentEl.style.position = 'relative';
                    //self._flash.setSize(400, 200); 
                    //vpaidParentEl.style.width = '100px';
                    //vpaidParentEl.style.height = '100px';
					//elementP.appendChild(vpaidParentEl);
					//elementP.style.zIndex=99999;
					//alert(elementP.id);
						document.getElementById("mycontoller").style.display="none";
				        adUnit.setAdVolume(0.1);
                        console.log('event:Start', err, result);
                        adUnit.startAd(function (err, result) {
                        console.log('startAd call', err, result);
                        });
               }
			});
			
            });


   console.log("принцип справедливости 1"); 
    if (!VPAIDFLASHClient.isSupported()) {
	    
         return onError('user don\'t support flash or doesn\'t have the minimum required version of flash ' + FLASH_VERSION);
    }
		

	    //this.el = swfobject.createSWF(swfConfig, params, this._flashID);

        if (!this.el) {
		    console.log("принцип справедливости 0"); 
            return onError( 'swfobject failed to create object in element' );
        }	
		

		
		
		 console.log("принцип справедливости 2 "); 
        function onError(error) {
            setTimeout(function(){
			    callback(new Error(error));
            }, 0);
            return me;
        }		
};
VPAIDFLASHClient.prototype._destroyAdUnit=function() {
         delete this._loadLater;

        if (this._adUnitLoad) {
            this._adUnitLoad = null;
            this._flash.removeCallback(this._adUnitLoad);
        }

        if (this._adUnit) {
            this._adUnit._destroy();
            this._adUnit = null;
        }
    }
VPAIDFLASHClient.prototype.destroy =function() {
        this._destroyAdUnit();

        if (this._flash) {
            this._flash.destroy();
            this._flash = null;
        }
        this.el = null;
        this._destroyed = true;
    }
VPAIDFLASHClient.prototype.loadAdUnit= function(adURL, callback) {

        if (this._flash.isReady()) {
		var self=this;
            this._adUnitLoad = function(err, message){
               if (!err) {
                   self._adUnit = new VPAIDAdUnit(self._flash);
                }
                self._adUnitLoad = null;
                callback(err, self._adUnit);
            };

         this._flash.callFlashMethod('loadAdUnit', [adURL], this._adUnitLoad);
        }else{

		 this._loadLater = {url: adURL, callback:callback};
		}

};


	function setStaticProperty(propertyName, value, writable) {
	    writable = writable || false;
        Object.defineProperty(VPAIDFLASHClient, propertyName, {
            writable: writable,
            configurable: false,
            value: value
        });
    }


	
	
setStaticProperty('isSupported', function(){
  
       //return true;
	   return swfobject.hasFlashPlayerVersion(FLASH_VERSION);
    }, true);
function hideFlashEl (el) {
    // can't use display none or visibility none because will block flash in some browsers
    el.style.position = 'absolute';
    el.style.left = '-1px';
    el.style.top = '-1px';
    el.style.width = '1px';
    el.style.height = '1px';
}

VPAIDFLASHClient.swfobject = swfobject;
module.exports =  VPAIDFLASHClient;
	 
/*






let flashTester = {isSupported: ()=> true}; // if the runFlashTest is not run the flashTester will always return true

class VPAIDFLASHClient {
    constructor (vpaidParentEl, callback, swfConfig = {data: 'VPAIDFlash.swf', width: 800, height: 400}, params = { wmode: 'transparent', salign: 'tl', align: 'left', allowScriptAccess: 'always', scale: 'noScale', allowFullScreen: 'true', quality: 'high'}, vpaidOptions = { debug: false, timeout: 10000 }) {

        var me = this;

        this._vpaidParentEl = vpaidParentEl;
        this._flashID = uniqueVPAID();
        this._destroyed = false;
        callback = callback || noop;

        swfConfig.width = isPositiveInt(swfConfig.width, 800);
        swfConfig.height = isPositiveInt(swfConfig.height, 400);

        createElementWithID(vpaidParentEl, this._flashID, true);

        params.movie = swfConfig.data;
        params.FlashVars = `flashid=${this._flashID}&handler=${JSFlashBridge.VPAID_FLASH_HANDLER}&debug=${vpaidOptions.debug}&salign=${params.salign}`;

        if (!VPAIDFLASHClient.isSupported()) {
            return onError('user don\'t support flash or doesn\'t have the minimum required version of flash ' + FLASH_VERSION);
        }

        this.el = swfobject.createSWF(swfConfig, params, this._flashID);

        if (!this.el) {
            return onError( 'swfobject failed to create object in element' );
        }

        var handler = callbackTimeout(vpaidOptions.timeout,
            (err, data) => {
                $loadPendedAdUnit.call(this);
                callback(err, data);
            }, () => {
                callback('vpaid flash load timeout ' + vpaidOptions.timeout);
            }
        );

        this._flash = new JSFlashBridge(this.el, swfConfig.data, this._flashID, swfConfig.width, swfConfig.height, handler);

        function onError(error) {
            setTimeout(() => {
                callback(new Error(error));
            }, 0);
            return me;
        }

    }

    destroy () {
        this._destroyAdUnit();

        if (this._flash) {
            this._flash.destroy();
            this._flash = null;
        }
        this.el = null;
        this._destroyed = true;
    }

    isDestroyed () {
        return this._destroyed;
    }

    _destroyAdUnit() {
        delete this._loadLater;

        if (this._adUnitLoad) {
            this._adUnitLoad = null;
            this._flash.removeCallback(this._adUnitLoad);
        }

        if (this._adUnit) {
            this._adUnit._destroy();
            this._adUnit = null;
        }
    }

    loadAdUnit(adURL, callback) {
        $throwIfDestroyed.call(this);

        if (this._adUnit) {
            this._destroyAdUnit();
        }

        if (this._flash.isReady()) {
            this._adUnitLoad = (err, message) => {
                if (!err) {
                    this._adUnit = new VPAIDAdUnit(this._flash);
                }
                this._adUnitLoad = null;
                callback(err, this._adUnit);
            };

            this._flash.callFlashMethod('loadAdUnit', [adURL], this._adUnitLoad);
        }else {
            this._loadLater = {url: adURL, callback};
        }
    }

    unloadAdUnit(callback = undefined) {
        $throwIfDestroyed.call(this);

        this._destroyAdUnit();
        this._flash.callFlashMethod('unloadAdUnit', [], callback);
    }
    getFlashID() {
        $throwIfDestroyed.call(this);
        return this._flash.getFlashID();
    }
    getFlashURL() {
        $throwIfDestroyed.call(this);
        return this._flash.getFlashURL();
    }
	
}

setStaticProperty('isSupported', () => {
    return swfobject.hasFlashPlayerVersion(FLASH_VERSION) && flashTester.isSupported();
}, true);

setStaticProperty('runFlashTest', (swfConfig) => {
    flashTester = createFlashTester(document.body, swfConfig);
});

function $throwIfDestroyed() {
    if(this._destroyed) {
        throw new Error('VPAIDFlashToJS is destroyed!');
    }
}

function $loadPendedAdUnit() {
    if (this._loadLater) {
        this.loadAdUnit(this._loadLater.url, this._loadLater.callback);
        delete this._loadLater;
    }
}

function setStaticProperty(propertyName, value, writable = false) {
    Object.defineProperty(VPAIDFLASHClient, propertyName, {
        writable: writable,
        configurable: false,
        value: value
    });
}


*/
